### PR TITLE
Query caching

### LINKED ISSUE #12: Cache search results

### PR DESCRIPTION
**Pull request by lothropgauss**
related to https://github.com/lbryio/lbry-sdk/issues/3393

### CODE REVIEW

**Comment by lothropgauss** [comment]
N.B. This works against the leveldb-resolve branch in lbry-sdk https://github.com/lbryio/lbry-sdk/tree/leveldb-resolve so deployment needs to wait until those changes are merged.

File: server/search.go [hunk 1]
```diff
@@ -179,43 +180,98 @@ func (s *Server) Search(ctx context.Context, in *pb.SearchRequest) (*pb.Outputs,
...
+	//		searchIndices = append(searchIndices, index)
+	//	}
+	//}
+
+	// If it's been more than RefreshDelta time since we last checked if the
+	// es index has been refreshed, we check (this is 2 seconds in prod,
+	// 0 seconds in debug / unit testing). If the index has been refreshed
+	// a different number of times since we last checked, we purge the cache
+	if time.Now().After(s.LastRefreshCheck.Add(s.RefreshDelta)) {
+		// FIXME: Should this be on all indices
```
Comment (lyoshenka) [inline server/search.go]: does this need fixing?

File: server/search.go [hunk 2]
```diff
@@ -179,43 +180,98 @@ func (s *Server) Search(ctx context.Context, in *pb.SearchRequest) (*pb.Outputs,
...
+		res, _ := client.IndexStats(searchIndices[0]).Do(ctx)
```
Comment (lyoshenka) [inline server/search.go]: i generally don't like swallowing errors. at the very least you should log it

File: server/search.go [hunk 3]
```diff
@@ -179,43 +180,98 @@ func (s *Server) Search(ctx context.Context, in *pb.SearchRequest) (*pb.Outputs,
...
+	//indices, _ := client.IndexNames()
+	//for _, index := range indices {
+	//	if index != "claims" {
+	//		log.Println(index)
+	//		searchIndices = append(searchIndices, index)
+	//	}
+	//}
+
+	// If it's been more than RefreshDelta time since we last checked if the
+	// es index has been refreshed, we check (this is 2 seconds in prod,
```
Comment (lyoshenka) [inline server/search.go:198]: ES data should only change if a new block comes in. Maybe in the future we can use that info to purge the cache so we avoid checking the index every 2 seconds

**Review by lyoshenka — APPROVED** [review]

File: server/search.go (hunk 3 above)
Comment (lothropgauss) [reply server/search.go:198]: This check shouldn't cost much, it's essentially just pinging the ES server, but we could definitely set the refresh delta higher to like 5 or maybe 10 seconds. @shyba we were talking about this over the weekend, any thoughts?

Comment (lyoshenka) [reply server/search.go:198]: this is fine. we can always bump it up later if its an issue. no need to worry about i tmuch now

File: server/search.go (hunk 1 above)
Comment (lothropgauss) [reply server/search.go]: No, on seconds though, I don't think so. We accept a single ES index as an argument, and in all our use cases I believe we only ever have to deal with one.

File: server/search.go (hunk 2 above)
Comment (lothropgauss) [reply server/search.go]: Done.

File: server/search.go (hunk 3 above)
Comment (lothropgauss) [reply server/search.go:198]: Cool, I parameterized it at least along with the cache ttl.

**Pull request merged**